doc/joinableTaskFactory.md

```diff
@@ -0,0 +1,117 @@
+In fact multiple processes with their own main threads can contribute to the token so that any of these processes that get called back as part of the request they themselves made, they will recognize the causal relationship and can mitigate deadlocks.
+
+### Processes with a main thread
+
+To resolve the deadlock, we need to follow the 3rd of [the threading rules of the `JoinableTaskFactory`](https://github.com/microsoft/vs-threading/blob/main/doc/threading_rules.md) by carefully applying `JoinableTaskFactory.RunAsync` such that the causal relationship is observed by the `JoinableTaskFactory` so it can mitigate the deadlocks.
+
+An application that has a main thread and an instance of `JoinableTaskContext` should set the `JsonRpc.JoinableTaskFactory` property to an instance of `JoinableTaskFactory`.
+Doing this has the following effects:
+
+1. An outbound JSON-RPC request that occurs within the context of a `JoinableTask` will include `joinableTaskToken` as a top-level property in the JSON-RPC message.
+  This property carries a token that represents the `JoinableTask` that needs the RPC call to complete.
+2. When an inbound JSON-RPC request carries a `joinableTaskToken` top-level property, the request will be dispatched within a `JoinableTask` that was created based on the token provided in the message.
+
+Taken together, these two effects ensure that when an RPC call requires the main thread to fulfill, and the RPC client is blocking the main thread, that the `JoinableTaskFactory` will be able to mitigate deadlocks by allowing the RPC server to access the main thread that is owned by the RPC client, assuming both parties are following the [`JoinableTaskFactory` threading rules](https://github.com/microsoft/vs-threading/blob/main/doc/threading_rules.md).
+
+This holds even when one or more intermediate processes exist between the client and server, provided each process propagates the `joinableTaskToken` top-level property.
+
+### Processes without a main thread
+
+Considering the more complex scenario in the above problem statement, there may be an intermediary process in a multi-hop RPC chain that itself doesn't have a main thread and thus no need for a `JoinableTaskFactory`.
+
+StreamJsonRpc makes accommodation for this scenario by automatically propagating the `joinableTaskToken` top-level property it sees in inbound requests to all outbound requests that come from that request.
+In other words, StreamJsonRpc considers that all outbound RPC requests are causally related to the inbound RPC request that caused the outbound request to be made.
+
+This does _not_ mean that any and all requests that are running concurrently with an outbound RPC request will be seen as causally related.
+Only the outbound requests that derive from the .NET execution context that dispatched an incoming request will be assigned that request's `joinableTaskToken`.
```

Comment:
I would love a more prescriptive description (something like "don't use `await Task.Run`"), or a short list of the most common practices that would take you off the current execution context.
Also, if I need to do any of the operations that would take you off the current execution context, like `await Task.Run`, is there a way to restore the causal relation (would using `JTF.RunAsync` instead of `Task.Run` help, even if I don't have a main thread myself)?

Reply:
`await Task.Run` wouldn't do it. In fact it's pretty far off the track before you can escape the ExecutionContext. You have to do something like `ThreadPool.UnsafeQueueUserWorkItem`. But this is beyond the scope of JSON-RPC specific concerns. If we needed to document this, it would belong in the vs-threading repo's docs. And so far, it hasn't come up. What _does_ come up are folks who occasionally _want_ to escape the ExecutionContext, and that's usually in connection with a problem with `AsyncLazy<T>` which also doesn't apply here.